File: product/server/dashboard/src/test/java/io/codekvast/dashboard/file_import/impl/PublicationImporterImplTest.java
package io.codekvast.dashboard.file_import.impl;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

import io.codekvast.common.lock.Lock;
import io.codekvast.common.lock.LockManager;
import io.codekvast.common.lock.LockTemplate;
import io.codekvast.common.messaging.CorrelationIdHolder;
import io.codekvast.dashboard.agent.AgentService;
import io.codekvast.dashboard.file_import.CodeBaseImporter;
import io.codekvast.dashboard.file_import.InvocationDataImporter;
import io.codekvast.dashboard.file_import.PublicationImporter;
import io.codekvast.dashboard.metrics.AgentMetricsService;
import io.codekvast.javaagent.model.v2.InvocationDataPublication2;
import io.codekvast.javaagent.model.v3.CodeBasePublication3;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.net.URISyntaxException;
import java.util.Collections;
import java.util.Optional;
import javax.validation.ConstraintViolation;
import javax.validation.Validator;
import org.assertj.core.util.Files;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.dao.DuplicateKeyException;

/** @author olle.hallin@crisp.se */
public class PublicationImporterImplTest {

  @Mock private CodeBaseImporter codeBaseImporter;

  @Mock private InvocationDataImporter invocationDataImporter;

  @Mock private Validator validator;

  @Mock private AgentMetricsService metricsService;

  @Mock private AgentService agentService;

  @Mock private LockManager lockManager;

  private PublicationImporter publicationImporter;

  @Before
  public void beforeTest() {
    MockitoAnnotations.initMocks(this);
    when(agentService.getCorrelationIdFromPublicationFile(any()))
        .thenReturn(CorrelationIdHolder.generateNew());
    this.publicationImporter =
        new PublicationImporterImpl(
            codeBaseImporter,
            invocationDataImporter,
            validator,
            metricsService,
            agentService,
            new LockTemplate(lockManager));
    when(lockManager.acquireLock(any()))
        .thenReturn(
            Optional.of(
                Lock.forPublication(
                    new File(
                        "/intake/queue/invocations-29-683ce793-8bb8-4dc9-a494-cb2543aa7964.ser"))));
  }

  @Test
  public void should_import_CodeBasePublication3() throws URISyntaxException {
    // given
    File file = new File(getClass().getResource("/sample-publications/codebase-v2.ser").toURI());
    when(codeBaseImporter.importPublication(any(CodeBasePublication3.class))).thenReturn(true);

    // when
    boolean handled = publicationImporter.importPublicationFile(file);

    // then
    assertThat(handled, is(true));

    verify(codeBaseImporter).importPublication(any(CodeBasePublication3.class));
    verify(validator).validate(any());
    verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator);
  }

  @Test
  public void should_reject_CodeBasePublication3_when_DuplicateKeyException()
      throws URISyntaxException {
    // given
    File file = new File(getClass().getResource("/sample-publications/codebase-v2.ser").toURI());
    when(codeBaseImporter.importPublication(any(CodeBasePublication3.class)))
        .thenThrow(new DuplicateKeyException("Thrown by mock"));

    // when
    boolean handled = publicationImporter.importPublicationFile(file);

    // then
    assertThat(handled, is(false));

    verify(codeBaseImporter).importPublication(any(CodeBasePublication3.class));
    verify(validator).validate(any());
    verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator);
  }

  @Test
  public void should_swallow_CodeBasePublication2_when_InvalidClassException()
      throws URISyntaxException {
    // given
    File file =
        new File(
            getClass()
                .getResource("/sample-publications/codebase-v2-bad-serialVersionUID.ser")
                .toURI());

    // when
    boolean handled = publicationImporter.importPublicationFile(file);

    // then
    assertThat(handled, is(true));

    verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator);
  }

  @Test
  public void should_import_InvocationDataPublication2() throws URISyntaxException {
    // given
    File file = new File(getClass().getResource("/sample-publications/invocations-v2.ser").toURI());
    when(invocationDataImporter.importPublication(any(InvocationDataPublication2.class)))
        .thenReturn(true);

    // when
    boolean handled = publicationImporter.importPublicationFile(file);

    // then
    assertThat(handled, is(true));
    verify(invocationDataImporter).importPublication(any(InvocationDataPublication2.class));
    verify(validator).validate(any());
    verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator);
  }

  @Test
  public void should_not_process_locked_file() throws URISyntaxException {
    // given
    File file = new File(getClass().getResource("/sample-publications/invocations-v2.ser").toURI());
    when(lockManager.acquireLock(any())).thenReturn(Optional.empty());

    // when
    boolean handled = publicationImporter.importPublicationFile(file);

    // then
    assertThat(handled, is(false));
    verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator);
  }

  @Test
  public void should_reject_InvocationDataPublication2_when_DuplicateKeyException()
      throws URISyntaxException {
    // given
    File file = new File(getClass().getResource("/sample-publications/invocations-v2.ser").toURI());
    when(invocationDataImporter.importPublication(any(InvocationDataPublication2.class)))
        .thenThrow(new DuplicateKeyException("Thrown by mock"));

    // when
    boolean handled = publicationImporter.importPublicationFile(file);

    // then
    assertThat(handled, is(false));
    verify(invocationDataImporter).importPublication(any(InvocationDataPublication2.class));
    verify(validator).validate(any());
    verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator);
  }

  @Test
  public void should_ignore_unrecognized_content() throws IOException {
    // given
    File file = Files.newTemporaryFile();
    file.deleteOnExit();

    try (ObjectOutputStream oos =
        new ObjectOutputStream(new BufferedOutputStream(new FileOutputStream(file)))) {
      oos.writeObject("Hello, World!");
    }

    // when
    boolean handled = publicationImporter.importPublicationFile(file);

    // then
    assertThat(handled, is(false));
    verify(validator).validate(anyString());
    verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator);
  }

  @SuppressWarnings("unchecked")
  @Test
  public void should_handle_invalid_content() throws URISyntaxException {
    // given
    File file = new File(getClass().getResource("/sample-publications/invocations-v2.ser").toURI());
    when(validator.validate(any()))
        .thenReturn(Collections.singleton(mock(ConstraintViolation.class)));

    // when
    boolean handled = publicationImporter.importPublicationFile(file);

    // then
    assertThat(handled, is(true));
    verify(validator).validate(any());
    verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator);
  }
}

File: product/server/intake/src/test/kotlin/io/codekvast/intake/file_import/impl/PublicationImporterImplTest.kt
package io.codekvast.intake.file_import.impl

import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import io.codekvast.common.lock.Lock
import io.codekvast.common.lock.LockManager
import io.codekvast.common.lock.LockTemplate
import io.codekvast.common.lock.LockTimeoutException
import io.codekvast.common.messaging.CorrelationIdHolder
import io.codekvast.intake.agent.service.AgentService
import io.codekvast.intake.file_import.CodeBaseImporter
import io.codekvast.intake.file_import.InvocationDataImporter
import io.codekvast.intake.file_import.PublicationImporter
import org.assertj.core.util.Files
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoMoreInteractions
import org.mockito.MockitoAnnotations
import org.springframework.dao.DuplicateKeyException
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.ObjectOutputStream
import java.util.*
import javax.validation.Validator
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/** @author olle.hallin@crisp.se
 */
internal class PublicationImporterImplTest {
    @Mock
    private lateinit var codeBaseImporter: CodeBaseImporter

    @Mock
    private lateinit var invocationDataImporter: InvocationDataImporter

    @Mock
    private lateinit var validator: Validator

    @Mock
    private lateinit var agentService: AgentService

    @Mock
    private lateinit var lockManager: LockManager

    private lateinit var publicationImporter: PublicationImporter

    @BeforeEach
    fun beforeTest() {
        MockitoAnnotations.openMocks(this)
        whenever(agentService.getCorrelationIdFromPublicationFile(any()))
            .thenReturn(CorrelationIdHolder.generateNew())
        whenever(lockManager.acquireLock(any()))
            .thenReturn(
                Optional.of(
                    Lock.forPublication(
                        File("/intake/queue/invocations-29-683ce793-8bb8-4dc9-a494-cb2543aa7964.ser")
                    )
                )
            )

        publicationImporter = PublicationImporterImpl(
            codeBaseImporter,
            invocationDataImporter,
            validator,
            agentService,
            LockTemplate(lockManager)
        )
    }

    @Test
    fun should_import_CodeBasePublication3() {
        // given
        val file = getResourceAsFile("/sample-publications/codebase-v2.ser")
        whenever(codeBaseImporter.importPublication(any()))
            .thenReturn(true)

        // when
        val handled = publicationImporter.importPublicationFile(file)

        // then
        assertTrue(handled)

        verify(codeBaseImporter).importPublication(any())
        verify(validator).validate(any<Any>())
        verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator)
    }

    @Test
    fun should_retry_CodeBasePublication3_when_DuplicateKeyException() {
        // given
        val file = getResourceAsFile("/sample-publications/codebase-v2.ser")
        whenever(codeBaseImporter.importPublication(any()))
            .thenThrow(DuplicateKeyException("Thrown by mock"))

        // when
        val handled = publicationImporter.importPublicationFile(file)

        // then
        assertFalse(handled)
        verify(codeBaseImporter).importPublication(any())
        verify(validator).validate(any<Any>())
        verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator)
    }

    @Test
    fun should_swallow_CodeBasePublication2_when_InvalidClassException() {
        // given
        val file = getResourceAsFile("/sample-publications/codebase-v2-bad-serialVersionUID.ser")

        // when
        val handled = publicationImporter.importPublicationFile(file)

        // then
        assertTrue(handled)
        verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator)
    }

    @Test
    fun should_import_InvocationDataPublication2() {
        // given
        val file = getResourceAsFile("/sample-publications/invocations-v2.ser")
        whenever(invocationDataImporter.importPublication(any()))
            .thenReturn(true)

        // when
        val handled = publicationImporter.importPublicationFile(file)

        // then
        assertTrue(handled)
        verify(invocationDataImporter).importPublication(any())
        verify(validator).validate(ArgumentMatchers.any<Any>())
        verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator)
    }

    @Test
    fun should_not_process_locked_file() {
        // given
        val file = getResourceAsFile("/sample-publications/invocations-v2.ser")
        whenever(lockManager.acquireLock(any())).thenReturn(Optional.empty())

        // when
        val handled = publicationImporter.importPublicationFile(file)

        // then
        assertFalse(handled)
        verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator)
    }

    @Test
    fun should_retry_InvocationDataPublication2_when_DuplicateKeyException() {
        // given
        val file = getResourceAsFile("/sample-publications/invocations-v2.ser")
        whenever(invocationDataImporter.importPublication(any()))
            .thenThrow(DuplicateKeyException("Thrown by mock"))

        // when
        val handled = publicationImporter.importPublicationFile(file)

        // then
        assertFalse(handled)
        verify(invocationDataImporter).importPublication(any())
        verify(validator).validate(any<Any>())
        verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator)
    }

    @Test
    fun should_retry_InvocationDataPublication2_when_LockTimeoutException() {
        // given
        val file = getResourceAsFile("/sample-publications/invocations-v2.ser")
        whenever(invocationDataImporter.importPublication(any()))
            .thenThrow(LockTimeoutException("Thrown by mock"))

        // when
        val handled = publicationImporter.importPublicationFile(file)

        // then
        assertFalse(handled)
        verify(invocationDataImporter).importPublication(any())
        verify(validator).validate(ArgumentMatchers.any<Any>())
        verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator)
    }

    @Test
    fun should_ignore_unrecognized_content() {
        // given
        val file = Files.newTemporaryFile()
        file.deleteOnExit()
        ObjectOutputStream(BufferedOutputStream(FileOutputStream(file))).use { oos ->
            oos.writeObject("Hello, World!")
        }

        // when
        val handled = publicationImporter.importPublicationFile(file)

        // then
        assertFalse(handled)
        verify(validator).validate(ArgumentMatchers.anyString())
        verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator)
    }

    @Test
    fun should_handle_invalid_content() {
        // given
        val file = getResourceAsFile("/sample-publications/invocations-v2.ser")
        whenever(validator.validate(any<Any>()))
            .thenReturn(setOf(mock()))

        // when
        val handled = publicationImporter.importPublicationFile(file)

        // then
        assertTrue(handled)
        verify(validator).validate(ArgumentMatchers.any<Any>())
        verifyNoMoreInteractions(codeBaseImporter, invocationDataImporter, validator)
    }

    private fun getResourceAsFile(path: String) = File(javaClass.getResource(path).toURI())
}
